### PR TITLE
strategy/periodic: support configuring a time zone for reboot windows

### PR DESCRIPTION
**Pull request by kelvinfan001**
Support configuring reboot windows defined by a user-specified time zone, instead of only UTC.

Include a single `time_zone` field (that defaults to UTC if not
specified) in the `periodic` strategy's configuration. Whenever we
need to check whether we're currently in an allowed reboot window,
we convert the current naive time to the naive time under the
specified time zone, and check whether we are currently in a reboot
window.

Under this implementation, we keep invariant clock time for reboots,
but do NOT keep invariant reboot window length; thus, in some cases,
reboot windows can be lengthened, shortened, or skipped entirely.

### CODE REVIEW

**Comment by kelvinfan001** [comment]
The current implementation uses Option 3 from https://github.com/coreos/zincati/issues/301#issuecomment-816728794.
This is still a work in progress as we need to properly address the third bullet point in https://github.com/coreos/zincati/issues/301#issuecomment-816685924.

File: src/strategy/mod.rs [hunk 1]
```diff
@@ -92,7 +92,13 @@ impl UpdateStrategy {
                 "{}, total schedule length {} minutes (next window {})",
                 self.configuration_label(),
                 p.schedule_length_minutes(),
-                p.human_remaining()
+                // If not using UTC, we cannot accurately calculate remaining time;
+                // display next window week day and time, instead.
+                if p.use_system_time {
+                    p.human_next_window()
```
Comment (kelvinfan001) [inline src/strategy/mod.rs:98]: `human_next_window()` attempts to return the week day and time of the next nearest reboot window. However, the current implementation _may_ lead to an incorrect next nearest reboot window if the function is called in the week of a "DST shift" (e.g. in Eastern Time, the week of "spring forward" or "fall back" day).

File: src/config/fragments.rs [hunk 2]
```diff
@@ -77,6 +77,9 @@ pub(crate) struct UpdateFleetLock {
 pub(crate) struct UpdatePeriodic {
     /// A weekly window.
     pub(crate) window: Option<Vec<UpdatePeriodicWindow>>,
+    /// Whether to use system time's time zone (i.e. `/etc/localtime`)
+    /// as the time zone for reboot windows.
+    pub(crate) use_system_time: Option<bool>,
```
Comment (lucab) [inline src/config/fragments.rs:82]: Bikeshedding: I think this may work better has a `use_utc: Option<bool>` which is on by default, and users have to explicitly opt-out.

File: src/strategy/periodic.rs [hunk 3]
```diff
@@ -2,19 +2,35 @@
...
+use tzfile::Tz;
 
 /// Strategy for periodic (weekly) updates.
-#[derive(Clone, Debug, Default, Serialize)]
+#[derive(Clone, Debug, Serialize)]
 pub(crate) struct StrategyPeriodic {
     /// Whitelisted time windows during which updates are allowed.
     schedule: WeeklyCalendar,
+    /// Time zone in which time windows are defined in.
+    #[serde(skip_serializing)]
```
Comment (kelvinfan001) [inline src/strategy/periodic.rs:23]: `tzfile::Tz` currently does not implement `Serialize`.

File: src/strategy/periodic.rs [hunk 4]
```diff
@@ -41,15 +58,57 @@ impl StrategyPeriodic {
...
+    fn get_time_zone(cfg: &inputs::PeriodicInput) -> Fallible<Tz> {
+        let tz;
+        if &cfg.time_zone == "localtime" {
+            // Use `read_link` instead of `exists()` because we only want to check for
+            // the existence of the `/etc/localtime` symlink, not whether it points to
+            // a valid location. (`read_link` returns an error if symlink doesn't exist)
+            if let Err(_) = read_link(Path::new("/etc/localtime")) {
+                tz = Tz::named("UTC");
+            } else {
+                tz = Tz::local();
```
Comment (kelvinfan001) [inline src/strategy/periodic.rs:78]: After switching to `anyhow`, we should probably add some context to the error that `Tz::local()` might return if `/etc/localtime` is a symlink to a bad/non-existent `tzfile`.
Currently, if `/etc/localtime` points to bogus, all we see is the following in the journal:
```
[ERROR] critical error: failed to assemble configuration settings
[ERROR]  -> failed to validate update-strategy configuration
[ERROR]  -> No such file or directory (os error 2)
```
It may be difficult to identify what went wrong in the update-strategy config.

File: src/strategy/periodic.rs [hunk 5]
```diff
@@ -41,15 +58,57 @@ impl StrategyPeriodic {
...
         Ok(strategy)
     }
 
+    /// Get the time zone from `periodic` strategy config.
+    fn get_time_zone(cfg: &inputs::PeriodicInput) -> Fallible<Tz> {
+        let tz;
+        if &cfg.time_zone == "localtime" {
+            // Use `read_link` instead of `exists()` because we only want to check for
+            // the existence of the `/etc/localtime` symlink, not whether it points to
+            // a valid location. (`read_link` returns an error if symlink doesn't exist)
```
Comment (kelvinfan001) [inline src/strategy/periodic.rs:74]: To add some context as to why I used `read_link()` here, we only want to default to `UTC` if `/etc/localtime` doesn't exist, and not if the location is invalid, in which case `Tz::local()` will return an Error. This is to prevent surprising users with a UTC time zone if they _think_ they're using `localtime`, but accidentally symlinked an invalid time zone when trying to configure their system's local time.

File: src/strategy/periodic.rs [hunk 6]
```diff
@@ -41,15 +58,57 @@ impl StrategyPeriodic {
...
+        let naive_utc_dt = Utc::now().naive_utc();
+        let dt = (&self.time_zone).from_utc_datetime(&naive_utc_dt);
+        let next_window_minute_in_week = self.schedule.next_window_minute_in_week(&dt);
+
+        match next_window_minute_in_week {
+            Some(minute_in_week) => {
+                let (weekday, hour, minute) = utils::weekly_minute_as_weekday_time(minute_in_week);
+                // `tmp_dt` is created just to use the `%Z` specifier of its `format()` method.
+                let tmp_dt = (&self.time_zone).from_utc_datetime(&Utc::now().naive_utc());
+                let tz_abbrev = tmp_dt.format("%Z").to_string();
```
Comment (kelvinfan001) [inline src/strategy/periodic.rs:102]: This is a hacky way to get an abbreviation or otherwise some sort of name of the configured time zone. Ideally, `tzfile::Tz` should implement `Display`.

**Review by kelvinfan001 — COMMENTED** [review]
I added some comments in places I feel like needs some extra attention :)

File: docs/usage/updates-strategy.md [hunk 7]
```diff
@@ -101,3 +101,35 @@ Reboot windows can be separately configured in multiple snippets, as long as eac
  * `length_minutes`: non-zero window duration, in minutes
 
 For convenience, multiple entries can be defined with overlapping times, and each window definition is allowed to cross day and week boundaries (wrapping to the next day).
+
```
Comment (lucab) [inline docs/usage/updates-strategy.md:104]: Let's add a `## Timezone configuration` header so that's it's easier to link to.

File: docs/usage/updates-strategy.md [hunk 8]
```diff
@@ -101,3 +101,35 @@ Reboot windows can be separately configured in multiple snippets, as long as eac
...
+To configure a non-UTC time zone for reboot windows, specify the `time_zone` field in a `updates.periodic` entry. The specified time zone must be either `"localtime"` or a time zone name from the [IANA Time Zone Database][IANA_tz_db] (you can find an unofficial list of time zone names [here][wikipedia_tz_names]). 
+If using `"localtime"`, `/etc/localtime` must either be a symlink to a valid `tzfile` entry in your system's local time zone database (under `/usr/share/zoneinfo/`), or not exist, in which case `UTC` is used.
+
+Note that you can only specify a single time zone for _all_ reboot windows.
+
+:warning: **When using non UTC time zones:** reboot window clock times are always obeyed, therefore **reboot window lengths may vary**!
+Reboot windows may be lengthened or shortened due to shifts in clock time. For example, with the `US/Eastern` time zone, on "fall back" day, a specified reboot window may be lengthened by up to one hour; on "spring forward" day, a specified reboot window may be shortened by up to one hour, or skipped entirely.
+
+Example of periodic reboot windows using the `US/Eastern` time zone.
```
Comment (lucab) [inline docs/usage/updates-strategy.md:113]: It may better to start with a simpler non-DST example first, and move this longer `US/Eastern` example and explanation to its own `## Timezone caveats` section.
`America/Panama` should be a good enough pick as it never introduced DST (so far) and is fairly recognizable by the average user.

File: src/config/fragments.rs [hunk 9]
```diff
@@ -77,6 +77,11 @@ pub(crate) struct UpdateFleetLock {
 pub(crate) struct UpdatePeriodic {
     /// A weekly window.
     pub(crate) window: Option<Vec<UpdatePeriodicWindow>>,
+    /// A time zone in the IANA Time Zone Database (https://www.iana.org/time-zones)
+    /// or "localtime".
```
Comment (lucab) [inline src/config/fragments.rs:81]: `If unset, UTC datetimes are used.`

File: src/strategy/mod.rs [hunk 10]
```diff
@@ -89,10 +89,16 @@ impl UpdateStrategy {
             UpdateStrategy::FleetLock(_) => self.configuration_label().to_string(),
             UpdateStrategy::Immediate(_) => self.configuration_label().to_string(),
             UpdateStrategy::Periodic(p) => format!(
-                "{}, total schedule length {} minutes (next window {})",
```
Comment (lucab) [inline src/strategy/mod.rs:92]: This `format!` logic became large enough that it can be moved to its own `p.calendar_summary()` or similar.

File: src/strategy/mod.rs [hunk 11]
```diff
@@ -89,10 +89,16 @@ impl UpdateStrategy {
...
+                "{}, total schedule length {} minutes; next window {}",
                 self.configuration_label(),
                 p.schedule_length_minutes(),
-                p.human_remaining()
+                // If not using UTC, we cannot accurately calculate remaining time;
+                // display next window week day and time, instead.
+                if p.time_zone != tzfile::Tz::named("UTC").unwrap() {
```
Comment (lucab) [inline src/strategy/mod.rs:97]: I think it makes sense to place this into a `p.is_calendar_utc() -> bool`, caching the boolean upfront instead of re-computing it each time.

File: src/strategy/mod.rs (hunk 1 above)
Comment (lucab) [reply src/strategy/mod.rs:98]: IMHO This is seem good enough already as you did. Do you think we should maybe consider doing `human_next_window()` unconditionally instead?

File: src/strategy/periodic.rs (hunk 3 above)
Comment (lucab) [reply src/strategy/periodic.rs:23]: From an out-of-band chat: we should consider adding `serde` support directly to that crate, possibly behind a feature gate.

File: src/strategy/periodic.rs (hunk 6 above)
Comment (lucab) [reply src/strategy/periodic.rs:102]: This is possibly something to bring to upstream, however it may be hard to sell as timezones don't have a canonical/preferred name (see for example `UTC` vs `Etc/UTC`).

File: src/strategy/periodic.rs [hunk 12]
```diff
@@ -41,15 +58,57 @@ impl StrategyPeriodic {
...
+                format!(
+                    "at {}:{} on {} ({}) if no shift in clock.",
```
Comment (lucab) [inline src/strategy/periodic.rs:104]: Up for your consideration, but you may turn the last part into a `(subject to timezone caveats)` if not-UTC.

File: src/strategy/mod.rs (hunk 1 above)
Comment (kelvinfan001) [reply src/strategy/mod.rs:98]: I think `human_next_window()` is easier to understand if the reader is in the same time zone as the configured time zone (likely the case when using a non-UTC time zone). But if the user is in a different time zone than the configured time zone (likely the case when using UTC), `human_next_window()` might require some careful mental gymnastics, especially if the reboot window is near.
TL;DR, I think `human_remaining()` is still useful if we're using UTC.

File: src/strategy/periodic.rs (hunk 6 above)
Comment (kelvinfan001) [reply src/strategy/periodic.rs:102]: Good point. Perhaps don't implement `Display`, but maybe have a `name()` method that returns the last bit of the path of the `tzfile(5)` that the `tzfile::Tz` struct reads its data from, e.g. two different `Tz` instances may return `America/Toronto` or `Canada/Eastern` even though the data should be the same.

File: src/strategy/periodic.rs (hunk 12 above)
Comment (kelvinfan001) [reply src/strategy/periodic.rs:104]: That's a good idea, since the next reboot window information isn't _always_ wrong even in the week of a shift in clock.

File: src/strategy/mod.rs (hunk 11 above)
Comment (kelvinfan001) [reply src/strategy/mod.rs]: We now have a `p.tz_name()` that gets the "cached" name of the time zone. Using this over a boolean because we need to use this name in `human_next_reboot_window()` too.

**Review by lucab — APPROVED** [review]

**Comment by jlebon** [comment]
Thanks, @kelvinfan001! Looking forward to using this on my FCOS server. :)

**Comment by travier** [comment]
Same here! Thanks!
